ark/internal/genai/team.go

```diff
@@ -86,33 +86,56 @@ func (t *Team) executeRoundRobin(ctx context.Context, userInput Message, history
 	messages := slices.Clone(history)
 	var newMessages []Message
 
-	for turn := 0; ; turn++ {
+	messageCount := 0 // Count individual agent messages
+	memberIndex := 0  // Track which agent should speak next
+
+	for {
 		// Check if context was cancelled
 		if ctx.Err() != nil {
 			return newMessages, ctx.Err()
 		}
 
-		turnTracker := NewExecutionRecorder(t.Recorder)
-		turnTracker.TeamTurn(ctx, "Start", t.FullName(), t.Strategy, turn)
+		// Check maxTurns before executing
+		if t.MaxTurns != nil && messageCount >= *t.MaxTurns {
+			turnTracker := NewExecutionRecorder(t.Recorder)
+			turnTracker.TeamTurn(ctx, "MaxTurns", t.FullName(), t.Strategy, messageCount)
+			// Log the maxTurns limit for observability, but return success with accumulated messages
+			// This preserves monitoring/alerting capabilities
+			t.Recorder.EmitEvent(ctx, "Warning", "TeamMaxTurnsReached", BaseEvent{
+				Name: t.FullName(),
+				Metadata: map[string]string{
+					"strategy":     t.Strategy,
+					"maxTurns":     fmt.Sprintf("%d", *t.MaxTurns),
+					"teamName":     t.FullName(),
+					"messageCount": fmt.Sprintf("%d", messageCount),
+				},
+			})
+			return newMessages, nil
+		}
 
-		for i, member := range t.Members {
-			// Check if context was cancelled before each member execution
-			if ctx.Err() != nil {
-				return newMessages, ctx.Err()
-			}
+		// Execute current agent
+		member := t.Members[memberIndex]
 
-			if err := t.executeMemberAndAccumulate(ctx, member, userInput, &messages, &newMessages, i); err != nil {
-				if IsTerminateTeam(err) {
-					return newMessages, nil
-				}
-				return newMessages, err
+		if err := t.executeMemberAndAccumulate(ctx, member, userInput, &messages, &newMessages, messageCount); err != nil {
+			if IsTerminateTeam(err) {
+				return newMessages, nil
 			}
-		}
 
-		if t.MaxTurns != nil && turn+1 >= *t.MaxTurns {
-			turnTracker.TeamTurn(ctx, "MaxTurns", t.FullName(), t.Strategy, turn+1)
-			return newMessages, fmt.Errorf("team round-robin MaxTurns reached %s", t.GetName())
+			// Log the error but continue with next agent (making ark resilient)
+			t.Recorder.EmitEvent(ctx, "Warning", "TeamMemberFailed", BaseEvent{
```

Comment:
This means that if we get a genuine error the query will show as completed right? I think if that's the case its better to fail the query to be safe and avoid masking errors